pkitesting/src/main/java/io/netty/pkitesting/RevocationServer.java

```diff
@@ -0,0 +1,194 @@
+ */
+public final class RevocationServer {
+    private static volatile RevocationServer instance;
+
+    private final HttpServer crlServer;
+    private final String crlBaseAddress;
+    private final AtomicInteger issuerCounter;
+    private final ConcurrentMap<X509Certificate, CrlInfo> issuers;
+    private final ConcurrentMap<String, CrlInfo> paths;
+
+    /**
+     * Get the shared revocation server instance.
+     * This will start the server, if it isn't already running, and bind it to a random port on the loopback address.
+     * @return The revocation server instance.
+     * @throws Exception If the server failed to start.
+     */
+    public static RevocationServer getInstance() throws Exception {
+        if (instance != null) {
+            return instance;
+        }
+        synchronized (RevocationServer.class) {
+            RevocationServer server = instance;
+            if (server == null) {
+                server = new RevocationServer();
+                server.start();
+                instance = server;
+            }
+            return server;
+        }
+    }
+
+    private RevocationServer() throws Exception {
+        // Use the JDK built-in HttpServer to avoid any circular dependencies with Netty itself.
+        crlServer = HttpServer.create(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0), 0);
+        crlBaseAddress = "http://localhost:" + crlServer.getAddress().getPort();
+        issuerCounter = new AtomicInteger();
+        issuers = new ConcurrentHashMap<>();
+        paths = new ConcurrentHashMap<>();
+        crlServer.createContext("/", exchange -> {
+            if ("GET".equals(exchange.getRequestMethod())) {
+                String path = exchange.getRequestURI().getPath();
+                CrlInfo info = paths.get(path);
+                if (info == null) {
+                    exchange.sendResponseHeaders(404, 0);
+                    exchange.close();
+                    return;
+                }
+                byte[] crl = generateCrl(info);
+                exchange.getResponseHeaders().put("Content-Type", Collections.singletonList("application/pkix-crl"));
+                exchange.sendResponseHeaders(200, crl.length);
+                try (OutputStream out = exchange.getResponseBody()) {
+                    out.write(crl);
+                    out.flush();
+                }
+            } else {
+                exchange.sendResponseHeaders(405, 0);
+            }
+            exchange.close();
+        });
+    }
+
+    private void start() {
+        if (Thread.currentThread().isDaemon()) {
+            crlServer.start();
+        } else {
+            // It's important the CRL server creates a daemon thread,
+            // because it's a singleton and won't be stopped except by terminating the JVM.
+            Thread th = new Thread(() -> crlServer.start());
+            th.setDaemon(true);
+            th.start();
+            try {
+                th.join();
```

Comment:
I don't understand this. You're still joining and blocking the non-daemon thread, what's the benefit of the daemon thread?

Reply:
Ahhh, HttpServer.start "inherits" the daemon-ness of the parent thread? From the code, I thought that it would be block forever and that's why you're joining. That makes more sense.

Reply:
I noticed that threads in the `ForkJoinPool.commonPool()` are always daemon, and JUnit 5 initializes the FJ common pool anyway. So I can let it call `start()` for me.